File: app/services/reputation_scorer.rb
# Computes a trust score (0-100) for a Nostr pubkey based on weighted signals.
#
# Design principles:
# - No single signal source can tank a score (caps per signal type)
# - Positive signals (friendship) offset negatives
# - Sensitivity is user-configurable (relaxed/moderate/strict)
# - Always reversible — scores inform auto-hide, never auto-ban
#
# Signals and base weights:
#   Your own hides of their messages:   -15 each (you trust yourself)
#   report_count from community:         -3 each (capped at -30, prevents brigading)
#   Banned from servers you're in:      -10 each (capped at -30)
#   Is your friend:                     +20 (trust boost)
#   Has been seen before (known):        +5
#
# Sensitivity multipliers (applied to penalties only):
#   relaxed:   0.5x  (tolerant, need strong evidence)
#   moderate:  1.0x  (balanced default)
#   strict:    1.5x  (cautious, low tolerance)
#
class ReputationScorer
  BASE_SCORE = 100

  SENSITIVITY_MULTIPLIERS = {
    "relaxed" => 0.5,
    "moderate" => 1.0,
    "strict" => 1.5
  }.freeze

  # Per-signal caps prevent any single signal type from dominating
  OWN_HIDE_WEIGHT = -15
  OWN_HIDE_CAP = -45 # max 3 hides before capped

  REPORT_WEIGHT = -3
  REPORT_CAP = -30 # 10 reports max effect

  BAN_WEIGHT = -10
  BAN_CAP = -30 # 3 bans max effect

  FRIEND_BONUS = 20
  KNOWN_BONUS = 5

  def initialize(pubkey, sensitivity: "moderate")
    @pubkey = pubkey
    @multiplier = SENSITIVITY_MULTIPLIERS[sensitivity] || 1.0
  end

  def score
    @score ||= compute_score
  end

  def breakdown
    @breakdown ||= compute_breakdown
  end

  private

  def compute_score
    bd = compute_breakdown
    raw = BASE_SCORE + bd[:penalties] + bd[:bonuses]
    raw.clamp(0, 100)
  end

  def compute_breakdown
    contact = Contact.find_by(pubkey: @pubkey)

    # --- Penalties ---

    # 1. Your own hides of their messages
    own_hides = Message.where(nostr_author_pubkey: @pubkey)
                       .where.not(hidden_at: nil)
                       .count
    own_hide_penalty = [ own_hides * OWN_HIDE_WEIGHT, OWN_HIDE_CAP ].max

    # 2. Community report count (aggregated from NIP-56 and other sources)
    report_count = contact&.report_count || 0
    # Also check remote members for additional reports
    remote_reports = RemoteMember.where(pubkey: @pubkey).maximum(:report_count) || 0
    total_reports = [ report_count, remote_reports ].max
    report_penalty = [ total_reports * REPORT_WEIGHT, REPORT_CAP ].max

    # 3. Banned from servers the local user is in
    ban_count = if User.first # single-user instance
      server_ids = User.first.servers.pluck(:id)
      if server_ids.any?
        # Check bans by matching pubkey to banned users' nostr keys
        banned_user_ids = User.where(nostr_public_key: @pubkey).pluck(:id)
        if banned_user_ids.any?
          Ban.where(server_id: server_ids, user_id: banned_user_ids).count
        else
          0
        end
      else
        0
      end
    else
      0
    end
    ban_penalty = [ ban_count * BAN_WEIGHT, BAN_CAP ].max

    # Apply sensitivity multiplier to all penalties
    total_penalties = ((own_hide_penalty + report_penalty + ban_penalty) * @multiplier).round

    # --- Bonuses (not affected by sensitivity) ---
    bonuses = 0
    bonuses += FRIEND_BONUS if contact&.accepted?
    bonuses += KNOWN_BONUS if contact.present?

    {
      own_hides: own_hides,
      own_hide_penalty: own_hide_penalty,
      report_count: total_reports,
      report_penalty: report_penalty,
      ban_count: ban_count,
      ban_penalty: ban_penalty,
      is_friend: contact&.accepted? || false,
      is_known: contact.present?,
      penalties: total_penalties,
      bonuses: bonuses
    }
  end
end

File: lib/services/reputation_scorer.dart
import 'dart:math';

import 'package:drift/drift.dart';
import '../database/database.dart';

/// Computes a trust score (0-100) for a Nostr pubkey based on weighted signals.
///
/// Signals and base weights:
///   Your own hides of their messages:   -15 each (cap -45)
///   report_count from community:        -3 each (cap -30)
///   Banned from servers you're in:      -10 each (cap -30)
///   Is your friend:                     +20
///   Has been seen before (known):       +5
///
/// Sensitivity multipliers (applied to penalties only):
///   relaxed:   0.5x
///   moderate:  1.0x
///   strict:    1.5x
class ReputationScorer {
  static const _baseScore = 100;

  static const _ownHideWeight = -15;
  static const _ownHideCap = -45;

  static const _reportWeight = -3;
  static const _reportCap = -30;

  static const _banWeight = -10;
  static const _banCap = -30;

  static const _friendBonus = 20;
  static const _knownBonus = 5;

  static const _sensitivityMultipliers = {
    'relaxed': 0.5,
    'moderate': 1.0,
    'strict': 1.5,
  };

  final InfernoDatabase _db;
  final String pubkey;
  final String sensitivity;

  ReputationScorer(this._db, this.pubkey, {this.sensitivity = 'moderate'});

  double get _multiplier => _sensitivityMultipliers[sensitivity] ?? 1.0;

  /// Compute the reputation score (0-100).
  Future<int> score() async {
    final bd = await breakdown();
    final raw = _baseScore + bd.penalties + bd.bonuses;
    return raw.clamp(0, 100);
  }

  /// Compute detailed breakdown of all signals.
  Future<ReputationBreakdown> breakdown() async {
    // Look up contact
    final contact = await (_db.select(_db.contacts)
          ..where((c) => c.pubkey.equals(pubkey)))
        .getSingleOrNull();

    // 1. Own hides of their messages
    final ownHides = await (_db.selectOnly(_db.messages)
          ..addColumns([_db.messages.id.count()])
          ..where(_db.messages.nostrAuthorPubkey.equals(pubkey) &
              _db.messages.hiddenAt.isNotNull()))
        .map((row) => row.read(_db.messages.id.count()) ?? 0)
        .getSingle();
    final ownHidePenalty = max(ownHides * _ownHideWeight, _ownHideCap);

    // 2. Community report count
    final contactReports = contact?.reportCount ?? 0;
    final remoteReportsQuery = await (_db.selectOnly(_db.remoteMembers)
          ..addColumns([_db.remoteMembers.reportCount.max()])
          ..where(_db.remoteMembers.pubkey.equals(pubkey)))
        .map((row) => row.read(_db.remoteMembers.reportCount.max()))
        .getSingleOrNull();
    final remoteReports = remoteReportsQuery ?? 0;
    final totalReports = max(contactReports, remoteReports);
    final reportPenalty = max(totalReports * _reportWeight, _reportCap);

    // 3. Bans from servers user is in
    final banCount = await _countBans();
    final banPenalty = max(banCount * _banWeight, _banCap);

    // Apply sensitivity multiplier to penalties
    final totalPenalties = ((ownHidePenalty + reportPenalty + banPenalty) * _multiplier).round();

    // Bonuses (not affected by sensitivity)
    int bonuses = 0;
    final isFriend = contact?.friendshipStatus == 3; // accepted
    if (isFriend) bonuses += _friendBonus;
    if (contact != null) bonuses += _knownBonus;

    return ReputationBreakdown(
      ownHides: ownHides,
      ownHidePenalty: ownHidePenalty,
      reportCount: totalReports,
      reportPenalty: reportPenalty,
      banCount: banCount,
      banPenalty: banPenalty,
      isFriend: isFriend,
      isKnown: contact != null,
      penalties: totalPenalties,
      bonuses: bonuses,
    );
  }

  Future<int> _countBans() async {
    // Bans table uses userId — look up user by nostrPublicKey first
    final bannedUser = await (_db.select(_db.users)
          ..where((u) => u.nostrPublicKey.equals(pubkey)))
        .getSingleOrNull();
    if (bannedUser == null) return 0;

    final bans = await (_db.selectOnly(_db.bans)
          ..addColumns([_db.bans.id.count()])
          ..where(_db.bans.userId.equals(bannedUser.id)))
        .map((row) => row.read(_db.bans.id.count()) ?? 0)
        .getSingle();
    return bans;
  }
}

class ReputationBreakdown {
  final int ownHides;
  final int ownHidePenalty;
  final int reportCount;
  final int reportPenalty;
  final int banCount;
  final int banPenalty;
  final bool isFriend;
  final bool isKnown;
  final int penalties;
  final int bonuses;

  const ReputationBreakdown({
    required this.ownHides,
    required this.ownHidePenalty,
    required this.reportCount,
    required this.reportPenalty,
    required this.banCount,
    required this.banPenalty,
    required this.isFriend,
    required this.isKnown,
    required this.penalties,
    required this.bonuses,
  });
}
